app/services/finance/ecf/output_calculator2.rb

```diff
@@ -0,0 +1,237 @@
+          fill_level = hash[:"#{declaration_type.underscore}_count"]
+
+          available = [fill_level, pot_size].min
+
+          hash[:"#{declaration_type.underscore}_count"] = fill_level - available
+          hash[:"#{declaration_type.underscore}_subtractions"] = available
+
+          pot_size -= available
+
+          hash
+        end
+
+        if pot_size.positive?
+          banding = banding.map do |hash|
+            available = hash[:"previous_#{declaration_type.underscore}_count"]
+
+            unless available.zero?
+              reduction = [pot_size, available].min
+
+              hash[:"#{declaration_type.underscore}_count"] = -reduction
+              hash[:"#{declaration_type.underscore}_subtractions"] += reduction
+
+              pot_size -= reduction
+            end
+
+            hash
+          end
+        end
+
+        banding.reverse
+      end
+
+      def previous_banding_for_declaration_type(declaration_type)
+        pot_size = previous_fill_level_for_declaration_type(declaration_type)
+
+        bands.zip(:a..:z).map do |band, letter|
+          band_capacity = band.max - (band.min || 1) + 1
+
+          fill_level = [pot_size, band_capacity].min
+
+          pot_size -= fill_level
+
+          key_name = "previous_#{declaration_type.underscore}_count".to_sym
+
+          {
+            band: letter,
+            min: band.min || 1,
+            max: band.max,
+            key_name => fill_level,
+          }
+        end
+      end
+
+      def bands
+        statement.contract.bands.order(max: :asc)
+      end
+
+      def band_letters
+        bands.zip(:a..:z).map { |e| e[1] }
+      end
+
+      def previous_fill_level_for_declaration_type(declaration_type)
+        billable = Finance::StatementLineItem
+          .where(statement: statement.previous_statements)
+          .billable
+          .joins(:participant_declaration)
+          .where(participant_declarations: { declaration_type: })
+          .count
+
+        refundable = Finance::StatementLineItem
+          .where(statement: statement.previous_statements)
+          .refundable
+          .joins(:participant_declaration)
+          .where(participant_declarations: { declaration_type: })
+          .count
+
+        billable - refundable
+      end
+
+      def previous_fill_level_for_uplift
+        billable = Finance::StatementLineItem
+          .where(statement: statement.previous_statements)
+          .billable
+          .joins(:participant_declaration)
+          .where("participant_declarations.sparsity_uplift = true OR participant_declarations.pupil_premium_uplift = true")
+          .count
+
+        refundable = Finance::StatementLineItem
+          .where(statement: statement.previous_statements)
+          .refundable
+          .joins(:participant_declaration)
+          .where("participant_declarations.sparsity_uplift = true OR participant_declarations.pupil_premium_uplift = true")
```

Comment:
Could this be made a bit clearer by merging the existing `ParticipantDeclaration.uplift` scope?

Reply:
yeah, could be dry-ed up a bit. the damage is fairly local so i'm gonna leave as is for now